feat: add custom html templates for components

keithgg opened the pull request:
## Description

At Opencraft, we have clients that want to be able to add custom HTML templates that show up while creating course components in Studio.

Currently, there's no way to achieve this besides forking `edx-platform` and adding the `yaml` files to `common/lib/xmodule/xmodule/templates/`

This PR allows operators to set a custom directory where templates can be additionally loaded from.

- Checks if `settings.CUSTOM_RESOURCE_TEMPLATES_DIRECTORY` exists.
- If so, load the `yaml` files with the same logic as in the `xmodule/templates` directory.

## Supporting information

- Configuration PR: https://github.com/openedx/configuration/pull/6706
- Custom Client Branch: https://github.com/open-craft/edx-platform/pull/453
- Task: https://tasks.opencraft.com/browse/BB-5885
- Sandbox: https://studio.pr30108.sandbox.opencraft.hosting [Sandbox]

## Testing instructions

- Update the setting `settings.CUSTOM_RESOURCE_TEMPLATES_DIRECTORY` to a valid path.
- In that directory create an `html` directory.
- Copy a template from `common/lib/xmodule/xmodule/templates/html` to the above directory with a different filename.
- In the Studio, edit a Unit. Under the **Text** component you should see your new template list (on the Sandbox I've added many).

## Deadline

"None"

## Other

**Settings**
```yaml
EDXAPP_CUSTOM_RESOURCE_TEMPLATES_DIRECTORY: /edx/var/edx-themes/edx-themes/resource-templates
edx_platform_commit: keith/bb_5583_themed_html_templates
edx_platform_repository_url: https://github.com/open-craft/edx-platform.git
configuration_source_repo_url: https://github.com/open-craft/configuration.git
configuration_version: keith/bb_5583_themed_html_templates
```

openedx-webhooks commented:
Thanks for the pull request, @keithgg!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

keithgg committed:
feat: added new setting CUSTOM_RESOURCE_TEMPLATES_DIRECTORY

This setting allows loading of Resource Templates from outside the
edx-platform codebase.

Operators will be able to add their own custom resource templates
without needing to fork the codebase.

natabene commented:
@keithgg Thank you for your contribution. Please let me know once it is ready for our review.

keithgg commented:
@natabene this is ready for review.

navinkarkera commented:
:+1:

- [x] I tested this: (followed steps mentioned in test instructions)
- [x] I read through the code
- [ ] I checked for accessibility issues
- [x] Includes documentation

giovannicimolin commented on cms/envs/common.py:
```diff
@@ -2098,6 +2098,12 @@
 
+# .. setting_name: CUSTOM_RESOURCE_TEMPLATES_DIRECTORY
+# .. setting_default: None
+# .. setting_description: Path to existing directory containing resource templates.
```
Just being nitpicky here: Can you add a bit more detail to the description? Looking from an outsiders perspective this is a somewhat cryptic variable that allows customizing something. What are the `resource templates` for? Who is using them? Are there any limitations?

keithgg replied:
Agreed. I'll add more detail to the descriptions.

giovannicimolin commented on lms/envs/common.py:
```diff
@@ -4413,6 +4413,12 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+# .. setting_description: Path to existing directory containing resource templates.
+#   "CUSTOM_RESOURCE_TEMPLATES_DIRECTORY" : null
+CUSTOM_RESOURCE_TEMPLATES_DIRECTORY = []
+
```
My understanding is that this is only ever needed in Studio, while creating content. 
Can you remove this from here? If not, can you explain why it needs to stay in a comment?

keithgg replied:
Hmmm, I think you're right. I thought at the time that the LMS needed to load the yaml files as well for display, but it might not be needed. I'll check and confirm.

keithgg replied:
It's needed. I'm not sure where, but without it the tests fail to pass, because modules on the LMS side use this code as well.

giovannicimolin commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -1055,21 +1085,21 @@ def templates(cls):
-                    template['template_id'] = template_file
-                    templates.append(template)
-        return templates
+                for template in cls._load_templates_in_dir(resource_filename(pkg, dirname)):
```
@keithgg Just me nitpicking: do you think we can rework this so that `get_template_dir` becomes `get_template_dirs` and returns a list of possible template directories so this can iterate over it?

Doing it like this would remove the need for code branches here and under `get_template` and make it easier to plug more directories in the future if we need to. What do you think?

keithgg replied:
Happy to do that. I left the existing `pkg_resources` way of loading to minimise my changes and because I couldn't quite figure out why it was using `pkg_resources` over an `os.listdir`.

keithgg replied:
Corrected.

giovannicimolin reviewed (changes requested):
@keithgg It's working perfectly for me. I left a few comments on the code, but mostly as suggestions to start a discussion.

Also, this needs a bit more documentation somewhere (maybe in the class description? and simple tips like file needs to be `yaml`, not `yml`, file structure needs to match the block type, etc).

keithgg commented on xmodule/x_module.py:
```diff
@@ -1626,7 +1687,6 @@ def hostname(self):
             'runtime.hostname is deprecated. Please use `LMS_BASE` from `django.conf.settings`.',
             DeprecationWarning, stacklevel=3,
         )
-        from django.conf import settings
```
This had to be removed, because of the settings import added at the top.

keithgg commented:
@giovannicimolin @pkulkark I've made the requested changes. Please take a look.

pkulkark commented on xmodule/x_module.py:
```diff
@@ -876,35 +877,64 @@ def policy_key(location):
+
+        for dirpath in cls.get_template_dirpaths():
+            for template in cls._load_templates_in_dir(dirpath):
+                templates[template['template_id']] = template
```
@keithgg Why use a dictionary object if in the end it only has to return a list (of the values)?

keithgg replied:
@pkulkark to remove duplicates. The calling code expects the `template_id` to be unique and this is to make sure you only ever have a single object corresponding to a specific `template_id`

pkulkark replied:
Ah I see. Thanks for the clarification!

pkulkark reviewed (approved):
@keithgg Just the one nit about using dictionary object over list. Otherwise LGTM :+1: 

:+1:

- [x] I tested this: Verified this works as expected (on a different sandbox)
- [x] I read through the code
- [ ] ~I checked for accessibility issues~ N/A
- [ ] ~Includes documentation~ N/A

pkulkark commented:
@giovannicimolin It looks like I can't merge this unless you add your approval too. Unless I don't have the right access to merge (I should have since I'm a core committer)? CC @natabene

giovannicimolin commented:
@pkulkark Approved based on your review.

openedx-webhooks commented:
@keithgg 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/30108&entry.1484655318=https://github.com/keithgg&entry.752974735=2022-03-23+11:24&entry.1917517419=2022-07-20+07:11&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
